### PR TITLE
Connect Wifi on High Power State

### PR DESCRIPTION
**Pull request by fustilio**
Integrate work done in PR https://github.com/bettr-xyz/tracestick-firmware/pull/39 into master and attempt to connect on startup and also on enter high power state

Output network status (i.e. connected or not) on menu, select to view status. Status currently doesn't update all the time but rather updates when a click is detected (as per original UI loop)

### CODE REVIEW

File: alpha/alpha.ino [hunk 1]
```diff
@@ -28,6 +28,13 @@ void setup() {
   // Start serial command console
   TS_SerialCmd.init();
   TS_SerialCmd.begin();
+
+  TS_RADIO.begin();
+
+  if (!TS_RADIO.wifi_is_connected()) {
+	  TS_RADIO.wifi_connect();
```
Comment (jamesi8086) [inline alpha/alpha.ino]: we don't want to connect on startup right? only in high power mode

Comment (jamesi8086) [reply alpha/alpha.ino]: also checks like `wifi_is_connected` are probably left to be within the wifi_connect function

File: alpha/power.cpp [hunk 2]
```diff
@@ -55,6 +56,7 @@ void _TS_POWER::lp_on_trans_hp()
 void _TS_POWER::hp_on_enter() {
   log_i("- Enter HP -");
   current_state = TS_PowerState::HIGH_POWER;
+  TS_RADIO.wifi_connect();
```
Comment (jamesi8086) [inline alpha/power.cpp]: how about when it leaves hp mode?

Comment (jamesi8086) [reply alpha/power.cpp]: maybe lets not put the wifi connect here but in the task with bt. if hp mode and not connected, try to connect every 1 minute (if not connected). That task would be in charge of anything radio related.

File: alpha/radio.cpp [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+//
+// Connect to WIFI
+// Scans for SSIDs, if preset SSID is found, attempt to connect, 
+// else prints the failed connection attempt. 
+void _TS_RADIO::wifi_connect()
+{
+  uint8_t tries = 0;
+  if (wifi_scan_networks())
+  {
+    while(!this->wifiConnected && tries < MAX_TRIES)
```
Comment (jamesi8086) [inline alpha/radio.cpp]: our retry loop is in the task, so lets not retry here

File: alpha/radio.cpp [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    {
+      //Attempt to connect to 'WIFI_SSID', prints name of connected WIFI. 
+      WiFi.begin(WIFI_SSID, WIFI_PASS);
+      // 2 second delay buffer to allow WiFi to connect 
+      TS_HAL.sleep(TS_SleepMode::Task, 2000);
```
Comment (jamesi8086) [inline alpha/radio.cpp]: task has other things to do, don't sleep here.

We can afford to check WiFi.status() multiple times when the task loops back to check if wifi is connected.

File: alpha/radio.cpp [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+  {
+    log_i("Known SSID not detected");
+  }
+}
+
+// disconnects from WIFI 
+void _TS_RADIO::wifi_disconnect()
+{
+  WiFi.disconnect();
+  TS_HAL.sleep(TS_SleepMode::Task, 1000);
```
Comment (jamesi8086) [inline alpha/radio.cpp]: again, we probably shouldn't have task sleeps here

File: alpha/radio.cpp [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+  }
+  else
+  {
+    log_i("Unsuccessful Disconnection from WIFI");
+  }
+}
+
+bool _TS_RADIO::wifi_is_connected()
+{
+  this->wifiConnected = WiFi.status() == WL_CONNECTED;
```
Comment (jamesi8086) [inline alpha/radio.cpp]: come to think of it, we already have WiFi.status()
do we still need wifiConnected ?

**Review by jamesi8086 — COMMENTED** [review]
I propose a structure where the task would periodically poll it to do stuff.
It appears that the wifi functions are mostly nonblocking.

wifi_enable(bool) - sets an internal flag on the desired status
wifi_update() - turns wifi on/off, scans, attempts connection, checks for connection, etc.

**Comment by zhenwei2972** [comment]
@fustilio are you going to address the changes or should I try? I think I'll need some pointers on how to put the retry loop in the task? 
Regarding this comment: "our retry loop is in the task, so lets not retry here" 
and to convert the task sleep to be asynch-compliant 
Regarding this comment: "async code, don't sleep here"

File: alpha/power.cpp (hunk 2 above)
Comment (fustilio) [reply alpha/power.cpp:59]: > @fustilio are you going to address the changes or should I try? I think I'll need some pointers on how to put the retry loop in the task?
> Regarding this comment: "our retry loop is in the task, so lets not retry here"
> and to convert the task sleep to be asynch-compliant
> Regarding this comment: "async code, don't sleep here"

I'll just quickly fix them up :p

**Comment by fustilio** [comment]
> @fustilio are you going to address the changes or should I try? I think I'll need some pointers on how to put the retry loop in the task?
> Regarding this comment: "our retry loop is in the task, so lets not retry here"
> and to convert the task sleep to be asynch-compliant
> Regarding this comment: "async code, don't sleep here"

No worries I'll just quickly fix them up :p

File: alpha/hal.h [hunk 7]
```diff
@@ -173,6 +173,7 @@ class _TS_HAL
 
   private:
     bool            bleInitialized;
+    bool            wifiInitialized;
```
Comment (jamesi8086) [inline alpha/hal.h]: do we still need this here?

Comment (fustilio) [reply alpha/hal.h]: I think no need, i'll remove it..

**Review by jamesi8086 — APPROVED** [review]

**Pull request merged**